nixos/modules/services/networking/prosody.nix

```diff
@@ -1,887 +1,682 @@
+            };
+            default = {};
+            description = mdDoc ''
+              Values specified here are applied to a specific virtual host and will override values set
+              in the global [settings](#opt-services.prosody.settings) option. Refer to <https://prosody.im/doc/configure>
+              for additional details.
+            '';
+          };
 
-        description = lib.mdDoc "Define the virtual hosts";
+          extraConfig = mkOption {
+            type = types.lines;
+            default = "";
+            description = mdDoc ''
+              Additional virtual host specific configuration.
+            '';
+          };
 
-        type = with types; attrsOf (submodule vHostOpts);
+          # options to preserve compatibility with this module pre nixos 23.11
 
-        example = {
-          myhost = {
-            domain = "my-xmpp-example-host.org";
-            enabled = true;
+          enabled = mkOption {
+            type = with types; nullOr bool;
+            default = null;
+            description = mdDoc "Whether to enable the virtual host.";
           };
-        };
 
-        default = {
-          localhost = {
-            domain = "localhost";
-            enabled = true;
+          ssl = mkOption {
+            type = types.nullOr (types.submodule {
+              options = {
+                key = mkOption {
+                  type = types.path;
+                  description = lib.mdDoc "Path to the key file.";
+                };
+
+                cert = mkOption {
+                  type = types.path;
+                  description = lib.mdDoc "Path to the certificate file.";
+                };
+
+                extraOptions = mkOption {
+                  type = types.attrs;
+                  default = {};
+                  description = lib.mdDoc "Extra SSL configuration options.";
+                };
+              };
+            });
+            default = null;
+            description = mdDoc "Paths to SSL files.";
           };
-        };
 
-      };
-
-      ssl = mkOption {
-        type = types.nullOr (types.submodule sslOpts);
-        default = null;
-        description = lib.mdDoc "Paths to SSL files";
-      };
+        };
 
-      admins = mkOption {
-        type = types.listOf types.str;
-        default = [];
-        example = [ "admin1@example.com" "admin2@example.com" ];
-        description = lib.mdDoc "List of administrators of the current host";
+        config.settings = {
+          ssl = mkMerge [
+            (mkIf (config.ssl != null) ({
+              key = config.ssl.key;
+              certificate = config.ssl.cert;
+            } // config.ssl.extraOptions))
+            (mkIf (config.useACMEHost != null) {
+              key = "${config'.security.acme.certs.${config.useACMEHost}.directory}/key.pem";
+              certificate = "${config'.security.acme.certs.${config.useACMEHost}.directory}/fullchain.pem";
+            })
+          ];
```

Comment:
With this variant option, the Direct TLS connection does not work.
Also need to add these lines:
```
  c2s_direct_tls_ssl = {
    key = "/var/lib/acme/example.com/key.pem";
    certificate = "/var/lib/acme/example.com/fullchain.pem";
  }
  s2s_direct_tls_ssl = {
    key = "/var/lib/acme/example.com/key.pem";
    certificate = "/var/lib/acme/example.com/fullchain.pem";
  }
```
And developers generally recommend importing certificates into `/etc/prosody/certs`.
Creating symlinks to the certificates in `/etc/prosody/certs` will probably work.

Reply:
> With this variant option, the Direct TLS connection does not work.

does that work in the current module? is this a regression or you're saying you want a new feature?


> And developers generally recommend importing certificates into `/etc/prosody/certs`.
> Creating symlinks to the certificates in `/etc/prosody/certs` will probably work.

if you put certs in `/etc/prosody/certs` then you don't need to specify an absolute path - but since we specify an absolute path anyways we don't need to do this.

Reply:
> does that work in the current module? is this a regression or you're saying you want a new feature?

The Direct TLS feature works if a single host is used and ssl certificates are specified in the global configuration.

> if you put certs in `/etc/prosody/certs` then you don't need to specify an absolute path - but since we specify an absolute path anyways we don't need to do this.

If you specify full access to certificates in the ssl block for the host, Direct TLS will stop working. Already checked.
To activate Direct TLS you will have to additionally specify parameters `c2s_direct_tls_ssl` and `s2s_direct_tls_ssl`.
Example:
```
VirtualHost "example.com"
  ssl = {
    key = "/var/lib/acme/example.com/key.pem";
    certificate = "/var/lib/acme/example.com/fullchain.pem";
  }
  c2s_direct_tls_ssl = {
    key = "/var/lib/acme/example.com/key.pem";
    certificate = "/var/lib/acme/example.com/fullchain.pem";
  }
  s2s_direct_tls_ssl = {
    key = "/var/lib/acme/example.com/key.pem";
    certificate = "/var/lib/acme/example.com/fullchain.pem";
  }
```